### PR TITLE
feat: Prep for v1

### LINKED ISSUE #7: Unable to Write to Stdin After Getting Output
**Describe the bug**
This means we cant:

1. assert url is (writes and reads output)
2. click a button

Or anything that tries to write after reading. See https://github.com/denoland/deno/issues/7727

**To Reproduce**
```
await dawn.assertUrlIs(...)
await dawn.click(...)
```

### PR DESCRIPTION
**Pull request by ebebbington**
Fixes #7 

**Description**

* Uses websockets instead of sub process to communicate

**Work**

- [x] Rewrite codebase to support web sockets instead
- [x] Tests
- [x] Make api scalable so it would be possible to use auth
- [x] Update docs
- [x] add website docs
- any others?

### CODE REVIEW

File: src/headless_browser.ts
Comment (caspervonb) [inline src/headless_browser.ts:23]: FYI; not all responses have an "inner" result, its common but not always.

Comment (ebebbington) [reply src/headless_browser.ts:23]: Ok cheers, gotcha 👍 thanks !

Comment (ebebbington) [reply src/headless_browser.ts]: https://github.com/drashland/dawn/pull/19/commits/44dbe7f41fdf810d8cdc1fcf81542b17d6fc2eab

**Comment by ebebbington** [comment]
This is the worst time i have ever had

**Comment by ebebbington** [comment]
ok well my recent code changes are ready for review, we can merge this PR, but cannot release yet as cinco/dawn is actually broken - i'll make an issue

**Comment by ebebbington** [comment]
FYI i made it to `HeadlessBrowser` is completely independent

**Comment by ebebbington** [comment]
Tests are 200% broken, they need to be refactored to reflect how the api works now

Issue is definitely the socket conn..  something isnt closing, even when using sanitise ops = false, the test process still runs

Im sure its down to the websocket bug

**Comment by ebebbington** [comment]
OMG THE TESTS PASS

**Comment by ebebbington** [comment]
ready for review now

File: src/headless_browser.ts
Comment (saragee3) [inline src/headless_browser.ts:175]: can we integrate `wocket`?

File: src/headless_browser.ts
Comment (saragee3) [inline src/headless_browser.ts:263]: similar to this... it could be helpful to have an `eval` function of something 

```
Sinco.eval(() => document.querySelector(...).children());
```

**Comment by saragee3** [comment]
this is cool... what's the advantage of using sockets vs sub process?

File: src/headless_browser.ts
Comment (ebebbington) [reply src/headless_browser.ts:175]: Wocket is a ws server right? Sinco just needs a ws client - or do you mean to use-wocket-client?

File: src/headless_browser.ts
Comment (ebebbington) [reply src/headless_browser.ts:263]: I agree, good shout! - i think this could be a separate issue to tackle though?

**Comment by ebebbington** [comment]
> this is cool... what's the advantage of using sockets vs sub process?

So i heard from @caspervonb that using websockets is the better approach when connecting to headless chromium session, plus, there's a bug with sub process that essentially stops sinco from working - both reasons were why i switched to using web sockets, plus, using a sub process, we'd be running the headless browser as a repl (`--repl`), meaning it could only act as how the dev tools console does (no checking if a page has loaded, solely JS expressions)

File: src/headless_browser.ts
Comment (saragee3) [reply src/headless_browser.ts:263]: agreed :)

File: src/headless_browser.ts
Comment (saragee3) [reply src/headless_browser.ts:175]: https://github.com/drashland/sinco/pull/19#issuecomment-739576066

same question i have for both of these comments... what's the best place i can read more about this?

Comment (ebebbington) [reply src/headless_browser.ts:175]: ill try get some info for you

**Review by saragee3 — APPROVED** [review]
i mean.... LGTM! 😎 I made a ticket for `eval` too. feel free to comment on the ticket regarding func name!

**Comment by ebebbington** [comment]
Woop woop! will wait for a review from @crookse or @Guergeiro then - after that... Sinco v1 here we come 😎

File: src/headless_browser.ts
Comment (Guergeiro) [inline src/headless_browser.ts:130]: Can we either:
1. Rewrite the entire `existsSync` function to be asynchronous?
2. In case that we use `existsSync` in a place that we cannot `await` it, maybe have an `exists` that is asynchronous and use it here?

File: src/utility.ts
Comment (Guergeiro) [inline src/utility.ts:15]: It's actually pretty easy to write *Promises* in Deno... I'm so used to Node that if I want to use a *Promise* I have to manually wrap their callbacks in a *Promise*. Well, everything in Deno that is not synchronous is already wrapped in a *Promise* 😀 
```ts
export const exists = async (filename: string): boolean => {
  try {
    await Deno.stat(filename);
    // successful, file or directory must exist
    return true;
  } catch (error) {
    if (error instanceof Deno.errors.NotFound) {
      // file or directory does not exist
      return false;
    } else {
      // unexpected error, maybe permissions, pass it along
      throw error;
    }
  }
};
```

**Review by Guergeiro — COMMENTED** [review]
Nitpicking here... Is just that I feel that if we are already in an `async` function, why not just use `await` everywhere we can instead of relying on synchronous jobs?

File: src/headless_browser.ts
Comment (ebebbington) [reply src/headless_browser.ts]: nitpicking is good, and good point, makes sense to use async: https://github.com/drashland/sinco/pull/19/commits/72e6566107044e5f8f0741cab9d2aa87bd076afd

File: src/utility.ts
Comment (ebebbington) [reply src/utility.ts:15]: same as above, makes sense to add a async version 👍 https://github.com/drashland/sinco/pull/19/commits/72e6566107044e5f8f0741cab9d2aa87bd076afd

**Review by Guergeiro — APPROVED** [review]

**Pull request merged**
